x/websocket/endpoint.go

```diff
@@ -155,6 +155,10 @@ func (c *gorillaConn) Read(buf []byte) (int, error) {
 		return 0, c.readErr
 	}
 	if err != nil {
+		var closeError *websocket.CloseError
+		if errors.As(err, &closeError) {
+			return 0, io.EOF
+		}
```

Comment:
We should report abnormal closes as errors, not EOF:
```suggestion
		if errors.As(err, &closeError) && closeError.Code == CloseNormalClosure {
			return 0, io.EOF
		}
```